vcf/src/main/java/org/pharmgkb/parsers/vcf/VcfDataParser.java

```diff
@@ -0,0 +1,167 @@
+			sf_logger.debug("Reading line #{}", m_lineNumber);
+		}
+
+		if (line.startsWith("#")) {
+			throw new BadDataFormatException("Line looks like metadata on line #" + m_lineNumber.get() + ": [[[" + line + "]]]");
+		}
+
+		List<String> data = sf_tab.splitToList(line);
+
+		try {
+
+			// CHROM
+			String chromosome = VcfEscapers.CHROMOSOME.unescape(data.get(0));
+
+			// POS
+			long position;
+			try {
+				position = Long.parseLong(data.get(1)) - 1; // VCF is 1-based
+			} catch (NumberFormatException e) {
+				throw new IllegalArgumentException("Position " + data.get(1) + " is not numerical");
+			}
+
+			// REF
+			String ref = data.get(3);
+
+			VcfPosition.Builder builder = new VcfPosition.Builder(chromosome, position, ref);
+
+			// ID
+			if (!data.get(2).equals(".")) {
+				builder.addIds(
+						sf_semicolon.splitToList(data.get(2)).stream()
+								.map(VcfEscapers.ID::unescape)
+								.collect(Collectors.toList())
+				);
+			}
+
+			// ALT
+			if (!data.get(4).equals(".")) {
+				builder.addAlts(sf_comma.splitToList(data.get(4)));
+			}
+
+			// QUAL
+			if (!data.get(5).equals(".")) {
+				try {
+					builder.setQuality(Optional.of(new GeneralizedBigDecimal(data.get(5))));
+				} catch (NumberFormatException e) {
+					throw new IllegalArgumentException("QUAL is not a number", e);
+				}
+			}
+
+			// FILTER
+			if (!data.get(6).equals(".")) {
+				Stream<String> unescaped = sf_semicolon.splitToList(data.get(6)).stream()
+						.map(VcfEscapers.FILTER::unescape);
+				builder.addFilters(unescaped.collect(Collectors.toList()));
+			}
+
+			// INFO
+			Set<String> keysUsed = new HashSet<>();
+			if (!data.get(7).equals(".")) {
+				List<String> props = sf_semicolon.splitToList(data.get(7));
+				for (String prop : props) {
+					int index = prop.indexOf('=');
+					String key, value;
+					if (index == -1) {
+						key = prop;
+						value = ""; // weird!
+					} else {
+						key = prop.substring(0, index);
+						value = prop.substring(index + 1);
+					}
+					if (keysUsed.contains(key)) {
+						throw new BadDataFormatException("Key " + key + " appears more than once in the INFO field");
+					}
+					Stream<String> unescaped = sf_comma.splitToList(value).stream()
+							.map(VcfEscapers.INFO_VALUE::unescape);
+					builder.putInfo(key, unescaped.collect(Collectors.toList()));
+					keysUsed.add(key);
+				}
+			}
+
+			// FORMAT
+			List<String> format = new ArrayList<>(0);
+			if (data.size() >= 9) {
+				format = sf_colon.splitToList(data.get(8));
+			}
+			builder.addFormats(format);
+
+			// samples
+			for (int x = 9; x < data.size(); x++) {
+				Stream<String> unescaped = sf_colon.splitToList(data.get(x)).stream()
+						.map(VcfEscapers.SAMPLE::unescape);
+				builder.addSample(new VcfSample.Builder(format, unescaped.collect(Collectors.toList())).build());
+			}
+
+			return builder.build();
+
+		} catch (IllegalArgumentException | ArrayIndexOutOfBoundsException e) {
```

Comment:
Why not throw this from the very beginning instead of catching and rethrowing?

It seems like you could simplify this into 2 portions, one that validated the input, and then one that worked on it.  Then it would also fail fast on errors.

Reply:
Totally agree. I think it was just easier.